tokio-trace-serde/src/lib.rs

```diff
@@ -0,0 +1,296 @@
+        }
+    }
+
+    fn record_i64(&mut self, field: &Field, value: i64) {
+        if self.state.is_ok() {
+            self.state = self.serializer
+                .serialize_entry(field.name(), &format_args!("{:?}", value))
+        }
+    }
+
+    fn record_str(&mut self, field: &Field, value: &str) {
+        if self.state.is_ok() {
+            self.state = self.serializer
+                .serialize_entry(field.name(), &format_args!("{:?}", value))
+        }
+    }
+}
+
+impl<S: SerializeMap> SerdeMapVisitor<S> {
+
+    /// Completes serializing the visited object, returning `Ok(())` if all
+    /// fields were serialized correctly, or `Error(S::Error)` if a field could
+    /// not be serialized.
+    fn finish(self) -> Result<S::Ok, S::Error> {
+        self.state?;
+        self.serializer.end()
+    }
+}
+
+struct SerdeStructVisitor<S: SerializeStruct> {
+    serializer: S,
+    state: Result<(), S::Error>
+}
+
+impl<S> Visit for SerdeStructVisitor<S> where S: SerializeStruct {
+    fn record_bool(&mut self, field: &Field, value: bool) {
+        // If previous fields serialized successfully, continue serializing,
+        // otherwise, short-circuit and do nothing.
+        if self.state.is_ok() {
+            self.state = self.serializer
+                .serialize_field(field.name(), &value)
+        }
+    }
+
+    fn record_debug(&mut self, field: &Field, value: &fmt::Debug) {
+        if self.state.is_ok() {
+            self.state = self.serializer
+                .serialize_field(field.name(), &format_args!("{:?}", value))
+        }
+    }
+
+    fn record_u64(&mut self, field: &Field, value: u64) {
+        if self.state.is_ok() {
+            self.state = self.serializer
+                .serialize_field(field.name(), &format_args!("{:?}", value))
+        }
+    }
+
+    fn record_i64(&mut self, field: &Field, value: i64) {
+        if self.state.is_ok() {
+            self.state = self.serializer
+                .serialize_field(field.name(), &format_args!("{:?}", value))
+        }
+    }
+
+    fn record_str(&mut self, field: &Field, value: &str) {
+        if self.state.is_ok() {
+            self.state = self.serializer
+                .serialize_field(field.name(), &format_args!("{:?}", value))
+        }
+    }
+}
+
+impl<S: SerializeStruct> SerdeStructVisitor<S> {
+
+    /// Completes serializing the visited object, returning `Ok(())` if all
+    /// fields were serialized correctly, or `Error(S::Error)` if a field could
+    /// not be serialized.
+    fn finish(self) -> Result<S::Ok, S::Error> {
+        self.state?;
+        self.serializer.end()
+    }
+}
+
+/// `AsSerde` is a trait that provides the `as_serde` function to types in
+/// `tokio-trace` to allow users to serialize their values.
+pub trait AsSerde<'a> {
+    type Serializable: serde::Serialize + 'a;
+
+    /// `as_serde` borrows a `tokio-trace` value and returns the serialized value.
```

Comment:
Technically this returns a serializ_able_ value --- a serde serializer is necessary to actually serialize the returned value.